### PR TITLE
fix(gemma4-mtp): make separate-model draft-mtp work (5 stacked fixes)

### LINKED ISSUE #108: Separate-model draft-mtp for Gemma 4 crashes on Windows/gfx1151 - 5 stacked bugs, root cause: n_embd_pre_norm width mismatch (1024 vs 5376)
### Environment

- Windows 11, Ryzen AI MAX+ 395 / Radeon 8060S (gfx1151), HIP SDK 7.1 **and** 7.2 — identical behavior on both
- ROCmFPX `main` (b9438-integration)
- Model: [kingjones777/Gemma-4-31B-it-ROCmFP4-GGUF](https://huggingface.co/kingjones777/Gemma-4-31B-it-ROCmFP4-GGUF) (`Q4_0_ROCMFP4_COHERENT` + `mtp-gemma-4-31B-it-Q8_0.gguf` draft)
- Repro command straight from the model card:

```
llama-server -m gemma-4-31B-it-Q4_0_ROCMFP4_COHERENT.gguf \
  --spec-type draft-mtp --model-draft mtp-gemma-4-31B-it-Q8_0.gguf \
  --spec-draft-ngl 999 --spec-draft-n-max 5 -ngl 999 -c 4096 -fa on -fit off
```

### Failure ladder (each bug masks the next)

1. **Draft context fails to create**: `Gemma4Assistant requires ctx_other to be set` — `server-context.cpp` deliberately does not set `ctx_other` for the separate-model `spec_mtp` branch, but the `GEMMA4_ASSISTANT` context ctor requires it.
2. After passing `ctx_other = ctx_tgt` for `LLM_ARCH_GEMMA4_ASSISTANT`: **GGML_ASSERT `tensor read out of bounds`** — `tensor read OOB: name=h_nextn type=f32 ne=[5376,0,1,1] nbytes=0 offset=0 size=57344`. On draft decodes that request no logits, `inp_out_ids` selects **zero rows**, and the last-layer `ggml_get_rows(inpL, inp_out_ids)` gives `h_next` zero rows while the extraction path reads `n_tokens × n_embd` floats.
3. After keeping a full-row residual for `h_next`: **KV positions not consecutive (`memory=17, batch=14`)** — `is_mem_shared` was detected as true for a **separate-model** draft (because `ctx_other == ctx_tgt` now holds), so all draft tokens were issued at the same position. A separate multi-head draft (n_mtp_layers=4) must use the `chain_heads` path.
4. After restricting `is_mem_shared` to same-model (embedded) MTP: **nondeterministic access violations** in `ggml_backend_tensor_set` (use-after-free) — fixed by PR #89 (`llm_graph_result::reset()` must also reset `t_h_pre_norm`).
5. **Root cause of the remaining AVs**: `llama_model::n_embd_pre_norm()` returns `hparams.n_embd` = **1024** for the gemma4-assistant draft (its internal MTP-layer width), but the h rows it consumes/produces (`inp_h`, `h_nextn`) are backbone-wide = **5376**. Consequences:
   - `llama_context::decode()` sizes the ubatch embd buffer with `model.n_embd_pre_norm()` → `udata->embd` is `n_tokens × 1024` floats;
   - `llm_graph_input_embd::set_input()` then reads `n_tokens × 5376` floats from that buffer → **OOB host read** (crash side confirmed in WinDbg: `db @rsi` → unmapped).

### Fix

All five are fixed in branch `fix/gemma4-mtp-windows` (commit 4625a54, PR to follow):

1. `tools/server/server-context.cpp`: set `cparams.ctx_other = ctx_tgt` when `model_dft->arch == LLM_ARCH_GEMMA4_ASSISTANT`.
2. `src/models/gemma4-assistant.cpp`: keep `inpL_full` (pre-`get_rows`) as the source for `h_next`.
3. `common/speculative.cpp`: `is_mem_shared` additionally requires `llama_get_model(ctx_tgt) == llama_get_model(ctx_dft)`; separate multi-head drafts fall back to `chain_heads`.
4. `src/llama-graph.cpp`: `t_h_pre_norm = nullptr` in `llm_graph_result::reset()` (same as PR #89) + `src/llama-context.cpp`: disable graph reuse for `ctx_other` pairs.
5. `src/llama-model.cpp`: `n_embd_pre_norm()` returns `hparams.n_embd_inp()` for `LLAMA_ARCH_GEMMA4_ASSISTANT`; MTP driver takes the h-row width from the **target**'s pre-norm width.

### Result

Stable, crash-free speculative decoding on Windows/gfx1151: **21–26 tok/s** on `Q4_0_ROCMFP4_COHERENT` + Q8 MTP draft (vs ~11 tok/s without the draft), coherent output, tool-calling verified over the OpenAI-compat API.

### PR DESCRIPTION
**Pull request by BEEugene**
﻿Fixes #108 (5 stacked bugs in the separate-model `draft-mtp` path for Gemma 4 on Windows/gfx1151; last commit of the ladder is the root cause вЂ” a 1024-vs-5376 h-row width mismatch).

## What

One commit, 7 files, all changes scoped to the gemma4 MTP path:

1. **`tools/server/server-context.cpp`** вЂ” pass `ctx_other = ctx_tgt` when the draft model arch is `LLAMA_ARCH_GEMMA4_ASSISTANT`. Without it the draft context ctor throws `Gemma4Assistant requires ctx_other to be set` and `llama-server`/`llama-cli` abort at load. (The existing NOTE about not setting `ctx_other` for separate-model MTP drafts concerns `is_mem_shared` detection, which is now guarded separately вЂ” see 3.)

2. **`src/models/gemma4-assistant.cpp`** вЂ” keep the full-row residual (`inpL_full`, before the last-layer `ggml_get_rows(inp_out_ids)`) as the source for `h_next`. Draft decodes request no logits, so `inp_out_ids` selects zero rows and `t_h_pre_norm` came out `[n_embd, 0]`; the extraction path then read `n_tokens Г— n_embd` floats out of it:
   `tensor read OOB: name=h_nextn ne=[5376,0,1,1] nbytes=0 ... size=57344`.

3. **`common/speculative.cpp`** вЂ” `is_mem_shared` now additionally requires `llama_get_model(ctx_tgt) == llama_get_model(ctx_dft)`. A **separate-model** multi-head draft (n_mtp_layers=4) must run the `chain_heads` path (one head per step, own KV + `seq_rm` rollback); treating it as shared-KV issued all draft tokens at the same position and later failed with `sequence positions ... Y = X + 1` (memory=17, batch=14). Also: the MTP driver's `n_embd` (h-row width for `batch.embd`) is taken from the **target**'s pre-norm width instead of the draft's internal `n_embd` (1024 vs 5376 вЂ” see 5).

4. **`src/llama-graph.cpp` / `src/llama-context.cpp`** вЂ” `llm_graph_result::reset()` also resets `t_h_pre_norm` (same one-liner as PR #89, credit @serrynaimo) plus graph reuse is disabled for `ctx_other` pairs; both eliminate nondeterministic use-after-free AVs after graph rebuilds.

5. **`src/llama-model.cpp`** вЂ” `llama_model::n_embd_pre_norm()` returns `hparams.n_embd_inp()` (backbone width, 5376) for `LLAMA_ARCH_GEMMA4_ASSISTANT`. Previously it returned `hparams.n_embd` = 1024 (the MTP-layer-internal width), so `llama_context::decode()` sized the ubatch embd buffer as `n_tokens Г— 1024` floats while `set_input()` read `n_tokens Г— 5376` from it вЂ” a host OOB read that showed up as access violations in `ggml_backend_tensor_set` (crash side verified in WinDbg: source pointer unmapped).

## Verification

Windows 11, Ryzen AI MAX+ 395 / 8060S (gfx1151), HIP SDK 7.2, model [kingjones777/Gemma-4-31B-it-ROCmFP4-GGUF](https://huggingface.co/kingjones777/Gemma-4-31B-it-ROCmFP4-GGUF) (`Q4_0_ROCMFP4_COHERENT` + `mtp-gemma-4-31B-it-Q8_0.gguf`), card-model command line (`-c 4096`, `--spec-draft-n-max 5`):

- 2+ full `llama-cli` runs, 100 tokens each: **21.8 / 22.8 / 25.7 tok/s**, coherent text, clean exit (no AV, no asserts).
- `llama-server` on `:1234`: OpenAI-compat chat + **tool-calling verified** (correct JSON function calls), sustained generation ~22 tok/s.
- Baseline without the draft: ~11 tok/s в†’ **~2Г— speedup**, matching the model card's numbers.
- Non-MTP paths unchanged (`Q4_0_ROCMFP4` without `--spec-*` still runs at ~11 tok/s; all five changes are gated on the gemma4-assistant/MTP paths).

### CODE REVIEW

**Comment by BEEugene** [comment]
Found the new repo here https://github.com/ROCmFPX/ROCmFPX - the bug was fixed there.
